### PR TITLE
(SERVER-2278) Add allow-subject-alt-names setting to puppetserver.conf

### PR DESCRIPTION
**Pull request by amysahli**
….conf

### CODE REVIEW

File: resources/ext/config/conf.d/puppetserver.conf [hunk 1]
```diff
@@ -74,3 +74,9 @@ profiler: {
     # enable or disable profiling for the Ruby code; defaults to 'true'.
     #enabled: true
 }
+
+# settings related to the certificate authority
+certificate-authority: {
+    # allow CA to sign certificate requests that have subject alternative names.
+    # allow-subject-alt-names: false
```
Comment (amysahli) [inline resources/ext/config/conf.d/puppetserver.conf]: Should I leave this commented out?

File: src/clj/puppetlabs/puppetserver/certificate_authority.clj [hunk 2]
```diff
@@ -110,6 +111,14 @@
 (def OIDMappings
   {schema/Str schema/Keyword})
 
+(def default-allow-subj-alt-names
+  false)
+
+(schema/defn ^:always-validate
+initialize-ca-config
+  [{:keys [allow-subject-alt-names] :or {allow-subject-alt-names default-allow-subj-alt-names} :as ca-data}]
```
Comment (justinstoller) [inline src/clj/puppetlabs/puppetserver/certificate_authority.clj]: I think the `:keys` and `:or` here are redundant.

File: resources/ext/config/conf.d/puppetserver.conf (hunk 1 above)
Comment (justinstoller) [reply resources/ext/config/conf.d/puppetserver.conf]: Yeah, having the defaults commented out at least matches the rest of the file

File: test/unit/puppetlabs/services/ca/certificate_authority_core_test.clj [hunk 3]
```diff
@@ -351,15 +351,28 @@
...
+     (testing "a CSR w/ DNS alt-names and disallowed subject-alt-names gets a specific error response"
+       (let [csr (io/input-stream (test-pem-file "hostwithaltnames.pem"))
+             settings (assoc settings :allow-subject-alt-names false)
+             response (handle-put-certificate-request!
+                       "hostwithaltnames" csr settings)]
+         (is (= 400 (:status response)))
+         (is (= (:body response)
+                (str "CSR 'hostwithaltnames' contains subject alternative names "
+                     "(DNS:altname1, DNS:altname2, DNS:altname3), which are disallowed. "
+                     "Use `puppet cert --allow-dns-alt-names sign hostwithaltnames` to sign this request.")))))
```
Comment (justinstoller) [inline test/unit/puppetlabs/services/ca/certificate_authority_core_test.clj]: We should update this error message while we're here

File: test/unit/puppetlabs/services/ca/certificate_authority_core_test.clj [hunk 4]
```diff
@@ -351,15 +351,28 @@
...
-      (testing "a CSR w/ DNS alt-names gets a specific error response"
-        (let [csr (io/input-stream (test-pem-file "hostwithaltnames.pem"))
-              response (handle-put-certificate-request!
-                        "hostwithaltnames" csr settings)]
-          (is (= 400 (:status response)))
-          (is (= (:body response)
-                 (str "CSR 'hostwithaltnames' contains subject alternative names "
-                      "(DNS:altname1, DNS:altname2, DNS:altname3), which are disallowed. "
-                      "Use `puppet cert --allow-dns-alt-names sign hostwithaltnames` to sign this request."))))))))
+     (testing "a CSR w/ DNS alt-names and disallowed subject-alt-names gets a specific error response"
```
Comment (justinstoller) [inline test/unit/puppetlabs/services/ca/certificate_authority_core_test.clj]: I don't think you need to add the `and disallowed subject-alt-names`, its the same thing as `DNS alt-names`, iiuc.

Comment (amysahli) [reply test/unit/puppetlabs/services/ca/certificate_authority_core_test.clj]: So the difference here is that the csr has dns alt-names AND the setting says not to allow them.

File: src/clj/puppetlabs/puppetserver/certificate_authority.clj [hunk 5]
```diff
@@ -130,6 +139,9 @@
   "The parent OID for all Puppet Labs specific X.509 certificate extensions."
   "1.3.6.1.4.1.34380.1")
 
+(def subject-alt-names-oid
```
[the remark this answers is not in the document]
Comment (amysahli) [reply src/clj/puppetlabs/puppetserver/certificate_authority.clj]: no problem

File: src/clj/puppetlabs/puppetserver/certificate_authority.clj (hunk 2 above)
Comment (amysahli) [reply src/clj/puppetlabs/puppetserver/certificate_authority.clj]: Is this not saying to get the allow-subject-alt-names key, and if it doesnt exist, to get the default allow-subject-alt-names?

File: test/unit/puppetlabs/services/ca/certificate_authority_core_test.clj (hunk 4 above)
Comment (justinstoller) [reply test/unit/puppetlabs/services/ca/certificate_authority_core_test.clj]: OOOooooooohhhhh

Comment (justinstoller) [reply test/unit/puppetlabs/services/ca/certificate_authority_core_test.clj]: Like allow-subject-alt-names settings if false

File: src/clj/puppetlabs/puppetserver/certificate_authority.clj (hunk 2 above)
Comment (justinstoller) [reply src/clj/puppetlabs/puppetserver/certificate_authority.clj]: Yeah, but isn't that basically what `(merge defaults ca-data)` does?

File: src/clj/puppetlabs/puppetserver/certificate_authority.clj [hunk 6]
```diff
@@ -110,6 +111,15 @@
...
 
+(def default-allow-subj-alt-names
+  false)
+
+(schema/defn ^:always-validate initialize-ca-config
+  "Adds in default ca config keys/values, which may be overwritten if a value for
+  any of those keys already exists in the ca-data"
+  [ca-data]
+  (let [defaults {:allow-subject-alt-names default-allow-subj-alt-names}]
+    (merge defaults ca-data)))
```
Comment (amysahli) [inline src/clj/puppetlabs/puppetserver/certificate_authority.clj]: @justinstoller This what you meant?

Comment (justinstoller) [reply src/clj/puppetlabs/puppetserver/certificate_authority.clj]: yep

**Comment by justinstoller** [comment]
Using the new CLI:
```
12:30:35 puppetserver-ca-cli justin (master $<>) ‖ be exe/puppetserver-ca list --all
Requested Certificates:
    bob   (SHA256)  61:E3:21:80:4B:2D:A6:4D:A1:73:5C:74:8D:3F:A6:FC:14:18:71:0F:43:B8:29:D9:3E:B3:22:CC:E2:56:64:36     alt names: ["DNS:bar", "DNS:baz", "DNS:bob", "DNS:foo"]
Signed Certificates:
    localhost   (SHA256)  D6:BB:5D:75:7C:D9:2D:C2:51:B4:F1:C4:07:C8:0D:55:E8:B5:2B:0F:9C:4C:20:C5:BF:0F:A0:13:DA:CF:E3:FD       alt names: ["DNS:puppet", "DNS:localhost"]

12:32:27 puppetserver-ca-cli justin (master $<>) ‖ be exe/puppetserver-ca sign --certname bob
Error:
    When download requested for #<struct Puppetserver::Ca::Utils::HttpClient::Result code="409", body="CSR 'bob' contains subject alternative names (DNS:bar, DNS:baz, DNS:bob, DNS:foo), which are disallowed. To allow subject alternative names, set allow-subject-alt-names to true in your puppetserver.conf file, restart the puppetserver, and try signing this certificate again.">
    code: 409
    body: CSR 'bob' contains subject alternative names (DNS:bar, DNS:baz, DNS:bob, DNS:foo), which are disallowed. To allow subject alternative names, set allow-subject-alt-names to true in your puppetserver.conf file, restart the puppetserver, and try signing this certificate again.
```
Updates dev/puppetserver.conf
```
12:36:17 puppetserver justin (merge-as *$) ‖ gd
diff --git a/dev/puppetserver.conf b/dev/puppetserver.conf
index 50ae9cb..11379eb 100644
--- a/dev/puppetserver.conf
+++ b/dev/puppetserver.conf
@@ -170,5 +170,5 @@ authorization: {
 }
 certificate-authority: {
-    allow-subject-alt-names: false
+    allow-subject-alt-names: true
 }
```

```
12:32:31 puppetserver-ca-cli justin (master $<>) ‖ be exe/puppetserver-ca list --all
Requested Certificates:
    bob   (SHA256)  61:E3:21:80:4B:2D:A6:4D:A1:73:5C:74:8D:3F:A6:FC:14:18:71:0F:43:B8:29:D9:3E:B3:22:CC:E2:56:64:36     alt names: ["DNS:bar", "DNS:baz", "DNS:bob", "DNS:foo"]
Signed Certificates:
    localhost   (SHA256)  D6:BB:5D:75:7C:D9:2D:C2:51:B4:F1:C4:07:C8:0D:55:E8:B5:2B:0F:9C:4C:20:C5:BF:0F:A0:13:DA:CF:E3:FD       alt names: ["DNS:puppet", "DNS:localhost"]

12:34:32 puppetserver-ca-cli justin (master $<>) ‖ be exe/puppetserver-ca sign --certname bob
Signed certificate for bob

12:34:36 puppetserver-ca-cli justin (master $<>) ‖ be exe/puppetserver-ca list --all
Signed Certificates:
    bob         (SHA256)  A3:AD:FF:69:FC:BA:3A:89:4F:2E:CB:46:29:CD:5C:B1:66:30:57:F9:CF:B7:6C:37:4A:E5:D9:AD:ED:13:F0:9A       alt names: ["DNS:bar", "DNS:baz", "DNS:bob", "DNS:foo"]
    localhost   (SHA256)  D6:BB:5D:75:7C:D9:2D:C2:51:B4:F1:C4:07:C8:0D:55:E8:B5:2B:0F:9C:4C:20:C5:BF:0F:A0:13:DA:CF:E3:FD       alt names: ["DNS:puppet", "DNS:localhost"]

```

**Pull request merged**